Support for tuples

jchavarri opened the pull request:
Fixes #68.

This was way easier than I expected, so I guess there must be many things missing 😄 I'll add some inline comments below.

Side note: I find the codebase to be really beautiful ✨  and I'm quite enjoying its discovery. What a delight to see how building on top of recursive types can lead to such a modular architecture!!

jchavarri commented on examples/reason-react-example/src/basics/Tuples.re:
```diff
@@ -0,0 +1,48 @@
+open Belt;
```
Should I add Tuples.re also in the TypeScript folder?

cristianoc replied:
See below.

jchavarri commented on src/Converter.re:
```diff
@@ -222,6 +226,8 @@ let rec converterIsIdentity = (~toJS, converter) =>
 
   | RecordC(_) => false
+  | TupleC(innerTypesC) =>
+    innerTypesC |> List.for_all(converterIsIdentity(~toJS))
```
Mostly the same as the [`ObjectC` case above](https://github.com/cristianoc/genType/pull/71/files#diff-b8ec785735c28040e738985bd4cce5e6L212), but didn't use the `OptionC` switch because I didn't think it was necessary (doesn't that case get process naturally when the recursion continues?)

cristianoc replied:
Option has no special role here, while it does for objects. So this is perfect.

jchavarri commented on src/Dependencies.re:
```diff
@@ -403,6 +403,18 @@ and translateTypeExpr_ =
+  | Ttuple(listExp) =>
+    let innerTypesTranslation =
+      listExp |> translateTypeExprs_(~language, ~typeVarsGen, ~typeEnv);
+    let innerTypes = innerTypesTranslation |> List.map(({typ, _}) => typ);
```
Very inspired on the `Tconstr(path, typeParams, _)` case [below](https://github.com/cristianoc/genType/pull/71/files#diff-d11ed9535d16ed2e430d872d7a86437bR445).

jchavarri commented on src/EmitTyp.re:
```diff
@@ -105,7 +105,14 @@ let rec renderTyp = (~language, ~indent=None, ~inFunType, typ) =>
+    "["
+    ++ (
+      innerTypes
+      |> List.map(typ => typ |> renderTyp(~language, ~indent, ~inFunType))
```
Should tuples be emitted like immutable arrays?

cristianoc replied:
It’s a good question. They’re immutable on the Reason side so I’d say yes.
Nit: typ => typ |> can be omitted: curry.

Also, TS has to is peculiar about naming types. Worth adding the examples. It might dislike the “[...]” for non-basic types.
Immutable arrays would get you out of that notation anyway.

jchavarri commented:
Thanks for the review! I have added the TS tests, and removed the unneeded lambda with `typ` param.

> > Should tuples be emitted like immutable arrays?

> They’re immutable on the Reason side so I’d say yes.

I have done a bit of research, and immutable tuples seem to have very bad support at the moment in both Flow and TypeScript.

In Flow, there is no way to mark a tuple or its inner types as immutable. `$ReadOnly` can only be used in objects, and adding `+` decorators only works for object fields. See related: https://github.com/facebook/flow/issues/4509.

In TypeScript, wrapping the tuple with `Readonly` seems to be valid syntactically, but it has very little effect in the final result:

```ts
let tuple: Readonly<[string, number]> = ['text', 3];

tuple[0] = 'new text'; // TS says: "All good!"
tuple.shift(); // Can remove elements from the tuple, even if it's read-only
let a: string = tuple[0]; // Type checks, but it'll be a number at runtime
```

[Playground](https://www.typescriptlang.org/play/index.html#src=let%20tuple%3A%20Readonly%3C%5Bstring%2C%20number%5D%3E%20%3D%20%5B'text'%2C%203%5D%3B%0D%0A%0D%0Atuple%5B0%5D%20%3D%20'new%20text'%3B%20%2F%2F%20TS%20says%3A%20%22All%20good!%22%0D%0Atuple.shift()%3B%20%2F%2F%20Can%20remove%20elements%20from%20the%20tuple%2C%20even%20if%20it's%20read-only%0D%0Alet%20a%3A%20string%20%3D%20tuple%5B0%5D%3B%20%2F%2F%20Type%20checks%2C%20but%20it'll%20be%20a%20number%20at%20runtime).

I don't really understand the purpose of `Readonly` in this case. 😕 

Considering this, I wonder if it's more appropriate to just leave tuples as they are, just to make more explicit they are mutable, to reflect better these limitations.

> can we just re-use everything done for the array type?

I thought about this. I'm not sure there are many things to reuse.

At the type definition level, TS syntax for tuples (`[one, two]`) is different than arrays (`string[]`), where the brackets go at the end. Flow syntax for arrays and tuples with brackets is the same, so in this case it could be reused. But considering how the Flow emitters for arrays are currently using the `$ReadOnlyArray<>` syntax, then this case would not allow for reusability either.

At the converter levels, arrays and tuples are different, with the first using `map` and the second being able to expand the array inline.

So in practice there seem to be different constraints for both cases, but I could explore trying to unify them more if you think it's worth it.

> TS has to is peculiar about naming types. Worth adding the examples. 

I added the `Tuple.re` tests to the TypeScript example, and I'm actually getting some errors, but I think they are not related to this addition.

The type `couple` [gets converted to an abstract class](https://github.com/cristianoc/genType/pull/71/files#diff-1bef4b1b86dec944cd55f6544030e039R28) and this causes a few issues:
- tslint warns that a class must be named with pascal case
- but more importantly, the functions that return or accept a value of type `couple` fail to type check because "Property 'opaque' is missing in type '{ name: any; age: any; }[]".

Do you have any pointers? I think in this case `couple` shouldn't be abstract / opaque 🤔 

Thanks again!

cristianoc commented:
Thanks for digging, this is great info! OK so it looks like immutability for tuples falls into the high investment low return category, and can be dropped.

`couple` sounds like the newly introduced tuple types are considered opaque. There's some pattern matching case analysis to decide what types are opaque, and this is probably a case of tuples missing from there.

cristianoc commented:
Wrong guess. You've handled the code for checking what's opaque.

cristianoc commented:
OK it was not far off. It's this line:
```
| Ident(_) => !(typ == booleanT || typ == numberT || typ == stringT)
```
meaning that any other compound of a tuple than those would be considered opaque.

So pre-existing.
Would you comment-out the example that triggers this. We can merge this, and I'll look into that separately.

jchavarri commented:
@cristianoc Done! 🙂

cristianoc commented:
So good. Thank you so much for this.
